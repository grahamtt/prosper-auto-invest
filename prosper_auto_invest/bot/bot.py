import argparse
import json
import logging
from collections import namedtuple
from datetime import datetime, timedelta, tzinfo

from humanize import naturaldelta
from time import sleep

import pytz
import requests

from prosper_api.auth_token_manager import AuthTokenManager
from prosper_api.client import Client
from prosper_api.config import Config
from prosper_api.models import AmountsByRating

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__file__)

TARGETS = {
    "NA": 0,
    "HR": 0.02,
    "E": 0.26,
    "D": 0.23,
    "C": 0.22,
    "B": 0.15,
    "A": 0.06,
    "AA": 0.06,
}

BucketDatum = namedtuple("BucketDatum", ["value", "pct_of_total", "error_pct"])


class Bot:
    def __init__(self):
        parser = argparse.ArgumentParser(
            prog="Prosper auto-invest",
            description="A bot that can find and invest in loans",
        )
        parser.add_argument(
            "-d",
            "--dry-run",
            help="Do everything but actually purchase the loans",
            action="store_true",
        )
        self.args = parser.parse_args()

        self.client = Client()

    def run(self):
        while True:
            account = self.client.get_account_info()
            logger.debug(json.dumps(account, indent=2))
            total_account_value = account.total_account_value
            buckets = {}
            invested_notes = account.invested_notes._asdict()
            pending_bids = account.pending_bids._asdict()
            for rating in invested_notes.keys():
                value = invested_notes[rating] + pending_bids[rating]
                pct_of_total = value / total_account_value
                buckets[rating] = BucketDatum(
                    value=value,
                    pct_of_total=pct_of_total,
                    error_pct=TARGETS[rating] - pct_of_total,
                )

            buckets["Cash"] = BucketDatum(
                account.available_cash_balance,
                account.available_cash_balance / total_account_value,
                0.0,
            )
            buckets["Pending deposit"] = BucketDatum(
                account.pending_deposit,
                account.pending_deposit / total_account_value,
                0.0,
            )
            buckets["Total value"] = BucketDatum(
                total_account_value, total_account_value / total_account_value, 0.0
            )

            cash = account.available_cash_balance
            grade_buckets_sorted_by_error_pct = sorted(
                buckets.items(), key=lambda v: v[1].error_pct, reverse=True
            )
            logger.info(f"Total value = ${total_account_value}")
            logger.info(
                f"Pending investments = ${account.pending_investments_primary_market:7.2f}"
            )
            for key, bucket in buckets.items():
                logger.info(
                    f"\t{key:16}= ${bucket.value:8.2f} ({bucket.pct_of_total * 100:6.2f}%) error: {bucket.error_pct*100:6.3f}%"
                )

            if cash >= 25 or self.args.dry_run:
                for target_grade, bucket in grade_buckets_sorted_by_error_pct:
                    logger.info(
                        f"Enough cash available; searching for something in {target_grade}"
                    )

                    listings = self.client.search_listings(
                        limit=500,
                        prosper_rating=[target_grade],
                        sort_by="lender_yield",
                        sort_dir="desc",
                    )

                    if not listings["result"]:
                        logger.info("No matching listings found")
                        continue
                    # listings['result'].sort(key=lambda v: v['historical_return'], reverse=True)
                    listing = listings["result"][0]
                    logger.debug(json.dumps(listing, indent=2))

                    invest_amount = 25 + cash % 25
                    lender_yield = listing["lender_yield"]
                    listing_number = listing["listing_number"]
                    if self.args.dry_run:
                        logger.info(
                            f"DRYRUN: Would have purchased ${invest_amount:5.2f} of listing {listing_number} at {lender_yield * 100:5.2f}%"
                        )
                    else:
                        order_result = self.client.order(listing_number, invest_amount)
                        logging.info(
                            f"Purchased ${invest_amount:5.2f} of {listing_number} at {lender_yield * 100:5.2f}%"
                        )
                        logging.debug(json.dumps(order_result, indent=2))
                    sleep_time_delta = timedelta(seconds=60)
                    break

            else:
                logger.info("Not enough cash available")
                now = datetime.now(pytz.timezone("America/Denver"))
                sleep_time_delta = (now + timedelta(days=1)).replace(
                    hour=7, minute=0, second=0, microsecond=0
                ) - now

            logger.info(f"Sleeping for {naturaldelta(sleep_time_delta)}")
            sleep(sleep_time_delta.total_seconds())

            # notes = client.list_notes(limit=100)
            # logger.info(json.dumps(notes, indent=2))
            #
            # listing = client.search_listings(listing_number=["13201299"], invested=True, biddable=False)
            # logger.info(json.dumps(listing, indent=2))


if __name__ == "__main__":
    Bot().run()
