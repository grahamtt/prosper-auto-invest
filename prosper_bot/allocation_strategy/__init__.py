from contextlib import AbstractContextManager
from datetime import datetime, timedelta
from typing import Callable, Iterable, Iterator, Optional

from prosper_api.client import Client
from prosper_api.models import Listing, SearchListingsRequest

__all__ = ["AllocationStrategy"]


def _lender_yield_sort(listing: Listing) -> bool:
    return listing.lender_yield


class AllocationStrategy(AbstractContextManager, Iterable[Listing]):
    """Defines a partial order over the set of active prosper listings.

    Specifically, it defines a sequence of prosper API calls to get listings and an optional sort method used to reorder them before emission. Use the timeout parameter to force a maximum time to wait for matching listings.
    """

    def __init__(
        self,
        client: Client,
        api_param_iterator: Iterator[SearchListingsRequest],
        local_sort: Optional[Callable[[Listing], bool]] = None,
        timeout_seconds: float = -1.0,
    ):
        """Gets an instance of AllocationStrategy.

        Args:
            client (Client): Prosper API client
            api_param_iterator (Iterator[SearchListingsRequest]): Iterates over the different parameters to pass the Listing api.
            local_sort (Optional[Callable[[Listing], bool]]): Sorts the returned listings into the desired final order before returning.
            timeout_seconds (float): The max real time in seconds to try to get a listing before giving up. `-1` (default) indicates not timeout is required.
        """
        if local_sort is None:
            local_sort = _lender_yield_sort
        self._client = client
        self._api_param_iterator = api_param_iterator
        self._local_sort = local_sort
        self._end_time = (
            datetime.now() + timedelta(seconds=timeout_seconds)
            if timeout_seconds > 0
            else None
        )
        self._params: Optional[SearchListingsRequest] = None
        self._buffer: Optional[Iterable[Listing]] = None

    def __next__(self) -> Listing:
        """Gets the next listing from the buffer or fetches a new one if needed and available.

        This method enforces the timeout set at instantiation.
        """
        if self._end_time is not None:
            if datetime.now() > self._end_time:
                raise StopIteration("Timed out while searching listings")

        if self._buffer is None:
            self._params = self._api_param_iterator.__next__()
            result = sorted(
                self._client.search_listings(self._params).result, key=self._local_sort
            )
            if len(result) == 0:
                raise StopIteration("Not more listings")
            self._buffer = iter(result)

        return next(self._buffer)

    def __iter__(self):
        """Implements the iterable interface."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Does nothing yet."""
        return None
