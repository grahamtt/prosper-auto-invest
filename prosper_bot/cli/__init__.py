from decimal import Decimal

from prosper_shared.omni_config import Config, input_schema
from schema import Optional


@input_schema
def _schema():
    return {
        Optional(
            "bot", default={"dry-run": False, "verbose": False, "min-bid": Decimal(25)}
        ): {
            Optional("dry-run", default=False): bool,
            Optional("verbose", default=False): bool,
            Optional("min-bid", default=Decimal(25.00)): Decimal,
            Optional("strategy", default="AGGRESSIVE"): str,
        }
    }


def build_config() -> Config:
    """Compiles all the config sources into a single config."""
    return Config.autoconfig(["prosper-api", "prosper-bot"], validate=True)
