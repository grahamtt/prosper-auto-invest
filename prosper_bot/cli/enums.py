from enum import Enum


class AllocationStrategies(Enum):
    """Provides a named collection of allocation strategies for use on the command line."""

    AGGRESSIVE = ""
