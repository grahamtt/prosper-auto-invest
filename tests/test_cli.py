import sys
from decimal import Decimal

import pytest

from prosper_bot.allocation_strategy import AllocationStrategies
from prosper_bot.bot import bot
from prosper_bot.cli import build_config


class TestCli:
    # Ensure the schema is loaded
    bot._schema()

    def test_build_config(self, mocker):
        mocker.patch.object(
            sys,
            "argv",
            [
                "test-cli",
                "--client-id=0123456789abcdef0123456789abcdef",
                "--username=fake-username",
                "--token-cache=fake-token-cache",
                "--dry-run",
                "--verbose",
                "--min-bid=30",
                "--target-loan-count=600"
                # TODO: the --strategy param is broken :(
                # "--strategy=CONSERVATIVE",
            ],
        )

        config = build_config()

        assert config._config_dict["prosper-bot"]["bot"] == pytest.approx(
            {
                "min-bid": Decimal("30"),
                "strategy": AllocationStrategies.AGGRESSIVE,
                "target-loan-count": 600,
            }
        )
        assert config._config_dict["prosper-bot"]["cli"] == {
            "dry-run": True,
            "verbose": True,
        }
        assert config._config_dict["prosper-api"] == {
            "auth": {"token-cache": "fake-token-cache"},
            "credentials": {
                "client-id": "0123456789abcdef0123456789abcdef",
                "username": "fake-username",
            },
        }
