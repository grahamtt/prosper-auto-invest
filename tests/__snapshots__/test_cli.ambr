# serializer version: 1
# name: TestCli.test_cli_help
  '''
  usage: prosper-bot [-h] [-c CLIENT-ID] [--client-secret CLIENT-SECRET]
                     [-u USERNAME] [-p PASSWORD] [-t TOKEN-CACHE] [-m MIN-BID]
                     [-s {AGGRESSIVE,CONSERVATIVE,OVERALL_HIGHEST_RATE}]
                     [--target-loan-count TARGET-LOAN-COUNT] [-v] [-d]
  
  options:
    -h, --help            show this help message and exit
  
  prosper-api.credentials:
    -c CLIENT-ID, --client-id CLIENT-ID
                          The client-id from Prosper; Type: str matching
                          /^[a-f0-9]{32}$/
    --client-secret CLIENT-SECRET
                          The client-secret from Prosper; can be stored and
                          accessed securely using the keyring library; Type: str
                          matching /^[a-f0-9]{32}$/
    -u USERNAME, --username USERNAME
                          Your Prosper username; Type: str
    -p PASSWORD, --password PASSWORD
                          Your Prosper password; can be stored and accessed
                          securely using the keyring library; Type: str
  
  prosper-api.auth:
    -t TOKEN-CACHE, --token-cache TOKEN-CACHE
                          The filesystem location where the auth token will be
                          cached; Type: str; Default:
                          /Users/graham/Library/Caches/prosper-api/token-cache
  
  prosper-bot.bot:
    -m MIN-BID, --min-bid MIN-BID
                          Minimum amount of a loan to purchase; Type: Decimal;
                          Default: 25.00
    -s {AGGRESSIVE,CONSERVATIVE,OVERALL_HIGHEST_RATE}, --strategy {AGGRESSIVE,CONSERVATIVE,OVERALL_HIGHEST_RATE}
                          Strategy for balancing your portfolio; Type: str;
                          Default: AGGRESSIVE
    --target-loan-count TARGET-LOAN-COUNT
                          Calculate min bid based on (total account value /
                          target loan count). Overrides min-bid if present;
                          Type: int
  
  prosper-bot.cli:
    -v, --verbose         Prints additional debug messages; Type: bool
    -d, --dry-run         Run the loop but don't actually place any orders;
                          Type: bool
  
  '''
# ---
